### PR TITLE
RDK-37042: Add interfaces for Airplay Plugin

### PR DESCRIPTION
**Pull request by tmanoj387**
Reason for change: Add wpeframework interfaces for airplay plugin RPC
Test Procedure: build success and JsonData_AirplayJSONRPC.h is created
Risks: None

Signed-off-by: Tanmay Shah <Tanmay_ManojkumarShah@comcast.com>

### CODE REVIEW

**Review by Srinuvasarao440 — APPROVED** [review]

File: jsonrpc/Airplay.json [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        "$ref": "#/common/results/void"
+      },
+      "errors": [
+        {
+          "description": "Airplay is unavailable",
+          "$ref": "#/common/errors/unavailable"
+        }
+      ]
+    },
+    "systemcommand": {
```
Comment (anand-ky) [inline jsonrpc/Airplay.json:72]: Pls elaborate on why this API is required.

File: jsonrpc/Airplay.json [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+          "description": "Airplay is unavailable",
+          "$ref": "#/common/errors/unavailable"
+        },
+        {
+          "description": "Command is not set",
+          "$ref": "#/common/errors/unknownkey"
+        }
+      ]
+    },
+    "factoryreset": {
```
Comment (anand-ky) [inline jsonrpc/Airplay.json:60]: Where does this factory reset happen? Does it reset the STB?

**Review by anand-ky — CHANGES_REQUESTED** [review]

File: jsonrpc/Airplay.json (hunk 2 above)
Comment (tmanoj387) [reply jsonrpc/Airplay.json:60]: There is Option under "Device Settings -> Factory reset". I believe user can choose to cleanup whole device (erase all sign-in info and app data) and restore device with original settings.

File: jsonrpc/Airplay.json (hunk 1 above)
Comment (tmanoj387) [reply jsonrpc/Airplay.json:72]: Ok I will add in new revision.

**Comment by tmanoj387** [comment]
@anand-ky, @Srinuvasarao440 

Patch 2 posted.

It is fixing Summary and description of systemcommand method.

File: interfaces/IAirplay.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                  virtual void Exit(const uint32_t exitCode) = 0;
+                  virtual void NotifyClient(const string& eventName) = 0;
+            };
+
+            virtual void Register(IAirplay::INotification* airplay) = 0;
+            virtual void Unregister(IAirplay::INotification* airplay) = 0;
+            virtual void FactoryReset() = 0;
+            virtual void SetVisible(bool visibility) = 0;
+            virtual void SystemCommand(const string& command) = 0;
+            virtual uint32_t startAirplayApp(const string& parametersJson) = 0;
```
Comment (pwielders) [inline interfaces/IAirplay.h:31]: Naming incorrectly, all Interface methods are using PascalCase. Also there some repetition in here. This is Part of the AirPlay interface so wht Repeat AirPlay in the name here. I guess a Proper name here would be *uint32_t StartApp (const string& parameters) = 0;*

File: interfaces/IAirplay.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+                PLAYING = 0x0001,
+                STOPPED = 0x0002,
+                SUSPENDING = 0x0004
+            };
+
+            struct EXTERNAL INotification : virtual public Core::IUnknown {
+                  enum { ID = ID_AIRPLAY_NOTIFICATION };
+                  virtual void StateChange(const IAirplay::state state) = 0;
+                  virtual void Exit(const uint32_t exitCode) = 0;
+                  virtual void NotifyClient(const string& eventName) = 0;
```
Comment (pwielders) [inline interfaces/IAirplay.h:23]: Repetition of naming in an INotification interface we do expect Notify...s to an interested party. Assumption is made here with the name that those interested must be Clients, are Servers not allowed to subscribe to these events ;-) , So would it not be more descriptive to say here _void Event(const string& eventName) = 0;_

File: interfaces/IAirplay.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+            };
+
+            virtual void Register(IAirplay::INotification* airplay) = 0;
+            virtual void Unregister(IAirplay::INotification* airplay) = 0;
+            virtual void FactoryReset() = 0;
+            virtual void SetVisible(bool visibility) = 0;
+            virtual void SystemCommand(const string& command) = 0;
```
Comment (pwielders) [inline interfaces/IAirplay.h:30]: As a rule of Thumb, we always make COMRPC (and thus interface methods that can pass a Process boundary) return an uint32_t to indicate if, for example, something went wrong during the communication. So preferably suggest to always return an uint32_t, If from the method perspective it should be void, just always return (Core::ERROR_NONE), if something during the communciation fails, the COM layer will change it into a value, indicating Communication error.

**Review by pwielders — CHANGES_REQUESTED** [review]
The JSONRPC interface resembled the COMRPC interface. Potentially you can heve the JSONRPC interface generated from the COMRPC interface by only adding the /* @JSON */ keyword. It will make future live and updates a bit easier :-)

File: interfaces/IAirplay.h (hunk 3 above)
Comment (tmanoj387) [reply interfaces/IAirplay.h:31]: I agree with fixing as PascalCase. I would like to fix this as StartAirPlayApp. This name is used in airplay-daemon to start Airplay application and just to sync with that name AirPlay was kept as it is. I hope it is fine.

File: interfaces/IAirplay.h (hunk 4 above)
Comment (tmanoj387) [reply interfaces/IAirplay.h:23]: Ack.

File: interfaces/IAirplay.h (hunk 5 above)
Comment (tmanoj387) [reply interfaces/IAirplay.h:30]: Thanks, I will use uint32_t for SystemCommand.

File: interfaces/IAirplay.h (hunk 3 above)
Comment (tmanoj387) [reply interfaces/IAirplay.h:31]: Actually I checked other .json files. Looks like it is preferred way to keep short names. I will keep "StartApp" and "StopApp". Thanks.

**Comment by tmanoj387** [comment]
@Srinuvasarao440, @anand-ky, @pwielders 

Thanks for reviews on Previous Patchset.
New Patchset is pushed for this PR with following changes:

- Add StopApp method
- Modify startAirplayApp name to StartApp
- Modify NotifyClient name to Event
- Change return type of SystemCommand from void to uint32_t

**Review by pwielders — APPROVED** [review]

**Review by anand-ky — APPROVED** [review]

**Pull request merged**